modules/createBrowserHistory.js

```diff
@@ -60,19 +60,15 @@ const createBrowserHistory = (props = {}) => {
     let path = pathname + search + hash
 
     warning(
-      !(basename && path.indexOf(basename) !== 0),
+      !(basename && hasBasename(path, basename)),
       'You are attempting to use a basename on a page whose URL path does not begin ' +
       'with the basename. Expected path "' + path + '" to begin with "' + basename + '".'
     )
 
     if (basename)
-      path = stripPrefix(path, basename)
+      path = stripBasename(path, basename)
 
-    return {
-      ...parsePath(path),
-      state,
-      key
-    }
+    return createLocation(path, state, key)
```

Comment:
I couldn't just default `pathname` to `/` in `parsePath` when it was an empty string  because that would break the resolving that is done using `history.location` I'm not altogether satisfied with this approach because now the hash history location objects will have a `key` property. Maybe this isn't a big deal, but it is a side effect of the change.